protocols/bftcosi/bftcosi.go

```diff
@@ -0,0 +1,615 @@
+	// if root we have finished
+	if bft.IsRoot() {
+		sig := bft.Signature()
+		if bft.onSignatureDone != nil {
+			bft.onSignatureDone(sig)
+		}
+		bft.Done()
+		return nil
+	}
+
+	// otherwise , send the response up
+	err := bft.SendTo(bft.Parent(), r)
+	bft.Done()
+	return err
+}
+
+func (bft *ProtocolBFTCoSi) handleResponsePrepare(r *Response) error {
+	// check if we have enough
+	bft.tprMut.Lock()
+	defer bft.tprMut.Unlock()
+	bft.tempPrepareResponse = append(bft.tempPrepareResponse, r.Response)
+	if len(bft.tempPrepareResponse) < len(bft.Children()) {
+		return nil
+	}
+
+	// wait for verification
+	bzrReturn, ok := bft.waitResponseVerification()
+	if ok {
+		// append response
+		resp, err := bft.prepare.Response(bft.tempPrepareResponse)
+		if err != nil {
+			return err
+		}
+		bzrReturn.Response = resp
+	}
+
+	dbg.Lvl3("BFTCoSi Handle Response PREPARE")
+	// if I'm root, we are finished, let's notify the "commit" round
+	if bft.IsRoot() {
+		bft.startChallengeCommit()
+
+		return nil
+	}
+	// send up
+	return bft.SendTo(bft.Parent(), bzrReturn)
+}
+
+// waitResponseVerification waits till  the end of the verification and returns
+// the BFTCoSiResponse along with the flag:
+// true => no exception, the verification is correct
+// false => exception, the verification is NOT correct
+func (bft *ProtocolBFTCoSi) waitResponseVerification() (*Response, bool) {
+	dbg.Lvl3("Waiting for response verification:", bft.Name())
+	r := &Response{
+		TYPE: RoundPrepare,
+	}
+	// wait the verification
+	verified := <-bft.verifyChan
+	if !verified {
+		// append our exception
+		r.Exceptions = append(r.Exceptions, cosi.Exception{
+			Public:     bft.Public(),
+			Commitment: bft.prepare.GetCommitment(),
+		})
+		dbg.Lvl3("Response verification: failed", bft.Name())
+		return r, false
+	}
+	dbg.Lvl3("Response verification: OK", bft.Name())
+
+	return r, true
+}
+
+// Signature will generate the final signature, the output of the BFTCoSi
+// protocol.
+func (bft *ProtocolBFTCoSi) Signature() *BFTSignature {
+	return &BFTSignature{
+		Sig:        bft.commit.Signature(),
+		Msg:        bft.Msg,
+		Exceptions: bft.tempExceptions,
+	}
+}
+
+// RegisterOnDone registers a callback to call when the bftcosi protocols has
+// really finished
+func (bft *ProtocolBFTCoSi) RegisterOnDone(fn func()) {
+	bft.onDoneCallback = fn
+}
+
+// RegisterOnSignatureDone register a callback to call when the bftcosi
+// protocol reached a signature on the block
+func (bft *ProtocolBFTCoSi) RegisterOnSignatureDone(fn func(*BFTSignature)) {
+	bft.onSignatureDone = fn
+}
+
+// nodeDone is either called by the end of EndProtocol or by the end of the
+// response phase of the commit round.
+func (bft *ProtocolBFTCoSi) nodeDone() bool {
+	dbg.Lvl3(bft.Name(), "nodeDone()")
+	bft.doneProcessing <- true
+	if bft.onDoneCallback != nil { // only true for the root
+		bft.onDoneCallback()
+	}
+	return true
+}
+
+const preparePrefix = 0x1
```

Comment:
Yes we can remove it, however in the prepare round we should sign the hash
of the bft.Msg not the actual Msg.

2016-04-15 17:20 GMT+02:00 Ismail Khoffi <notifications@github.com>:

> In protocols/bftcosi/bftcosi.go
> <https://github.com/dedis/cothority/pull/354#discussion_r59891127>:
>
> > +func (bft *ProtocolBFTCoSi) RegisterOnSignatureDone(fn func(*BFTSignature)) {
> > +	bft.onSignatureDone = fn
> > +}
> > +
> > +// nodeDone is either called by the end of EndProtocol or by the end of the
> > +// response phase of the commit round.
> > +func (bft *ProtocolBFTCoSi) nodeDone() bool {
> > +	dbg.Lvl3(bft.Name(), "nodeDone()")
> > +	bft.doneProcessing <- true
> > +	if bft.onDoneCallback != nil { // only true for the root
> > +		bft.onDoneCallback()
> > +	}
> > +	return true
> > +}
> > +
> > +const preparePrefix = 0x1
>
> It was there to differentiate between the phases. It's not necessary
> anymore. @LefKok <https://github.com/LefKok>: Are you OK, with removing
> it completely?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/dedis/cothority/pull/354/files/f7e372d4f0f0ecddf30556b3495f5b540096bffc#r59891127>
>



-- 

Kokoris-Kogias Eleftherios
PhD Candidate at School of  Computer  and Communication Sciences
EPFL